### PR TITLE
ELEC-19: Event queue data structures

### PR DESCRIPTION
**Pull request by tituschow**
Added priority queue, object pool, and event queue data structures for use with the FSMs and their associated unit tests.

### CODE REVIEW

File: inc/event_queue.h [hunk 1]
```diff
@@ -1,10 +1,12 @@
 #pragma once
 #include <stdint.h>
 #include <stdbool.h>
+#include "objpool.h"
```
Comment (ckitagawa) [inline inc/event_queue.h:7]: I'd like to propose system headers and regular headers should have a newline between them, same with the pragma guards and system headers. This follows the standard Google C++ include rules. Also headers should be in alphabetical order just makes it easier for readability when lots of includes are present. If you are okay with this I will add it to the style guide and linter.

File: inc/misc.h [hunk 2]
```diff
@@ -0,0 +1,4 @@
+#pragma once
+// Common helper macros
+
+#define SIZEOF_ARRAY(arr) (sizeof((arr))/sizeof((arr)[0]))
```
Comment (ckitagawa) [inline inc/misc.h:4]: Glad this file exists :) A more descriptive name like helpers or macros might be good.

File: inc/objpool.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+// Object Pool Interface
+// Manages a pre-allocated array of objects. We use this instead of a heap so we don't need to deal
+// with memory fragmentation. Use designated initializers to prevent losing object marker information.
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include "misc.h"
```
Comment (ckitagawa) [inline inc/objpool.h:9]: Same thing as earlier about headers assuming you agree with it.

File: inc/pqueue.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+// Generic priority queue
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
```
Comment (ckitagawa) [inline inc/pqueue.h:5]: Again about headers.

File: src/event_queue.c [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// Event queue uses a priority heap and object pool under the hood.
+// Currently, there is only one global event queue.
+#include "event_queue.h"
+#include "pqueue.h"
+#include "objpool.h"
```
Comment (ckitagawa) [inline src/event_queue.c:5]: Headers order.

File: src/objpool.c [hunk 6]
```diff
@@ -0,0 +1,57 @@
+// The object pool uses a flag and index to verify a given node and determine its state.
+// We currently use a bitfield to maximize the index limit.
+#include <string.h>
+#include <stdio.h>
+#include "objpool.h"
```
Comment (ckitagawa) [inline src/objpool.c:4]: Again about headers.

File: inc/objpool.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#pragma once
```
Comment (ckitagawa) [inline inc/objpool.h:1]: This objpool is brilliant! I might make a map with it.

File: src/pqueue.c [hunk 8]
```diff
@@ -0,0 +1,67 @@
+// Implements a min-heap.
+// Modified from http://cs.smu.ca/~pawan/teach/csc342/98-99-2/book/chapter9/minheap.h
+#include <stdlib.h>
+#include <string.h>
+#include "pqueue.h"
```
Comment (ckitagawa) [inline src/pqueue.c:5]: Headers.

File: src/pqueue.c [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+void pqueue_init(PQueue *queue, PQueueNode *nodes, size_t num_nodes) {
+  memset(queue, 0, sizeof(*queue));
+
+  queue->nodes = nodes;
+  queue->max_nodes = num_nodes - 1;  // 1-indexed heap - throw away one node
+}
+
+bool pqueue_push(PQueue *queue, void *data, uint16_t prio) {
+  if (queue->size == queue->max_nodes) {
+    // Ran out of space - TODO: handle errors better
```
Comment (ckitagawa) [inline src/pqueue.c:16]: Want me to write a status library that allows for more verbose error handling without exceptions?

ie :
status = 0 = OK
status > 0 = SOME_ENUMERATED_ERROR (optionally with fixed char message).
These could be accompanied by a macro/function that will dump the error to CAN/USB/UART for debugging. Also it allows for some more flexibility with handling multiple failure states from a single function ie if the queue is full then try again after a backoff?

File: test/test_objpool.c [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  };
+
+  // Copy the data from the stack object. This will result in an invalid marker.
+  TestObject *node = objpool_get_node(&gv_pool);
+  TEST_ASSERT_NOT_NULL(node);
+  *node = data;
+
+  // We should still be able to free the node if this happens.
+  TEST_ASSERT_TRUE(objpool_free_node(&gv_pool, node));
+}
```
Comment (ckitagawa) [inline test/test_objpool.c]: No newline at EOF (even GitHub gave you a linter error in the review :p)

File: inc/pqueue.h [hunk 11]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+// Generic priority queue
```
Comment (ckitagawa) [inline inc/pqueue.h]: I know you mention it is min later but you should probably state if it is a min or max pqueue in the header description.

**Review by ckitagawa — CHANGES_REQUESTED** [review]
A few nitpicks and points to discuss. Very cool implementation of the objpool. I'm going to trust your tests will catch any issues with the heap's behavior for the pqueue/event queue.

File: inc/event_queue.h (hunk 1 above)
Comment (tituschow) [reply inc/event_queue.h:7]: I mean, it doesn't really matter to me tbh. If it gets fixed automatically by the linter I'm okay with it.
We should probably define a rule for module documentation and where it should go.

File: src/pqueue.c (hunk 9 above)
Comment (tituschow) [reply src/pqueue.c:16]: Yeah, error handling is always a tricky problem. We should probably add one.

File: test/test_objpool.c (hunk 10 above)
Comment (tituschow) [reply test/test_objpool.c]: whoops :)

**Review by ckitagawa — APPROVED** [review]

**Pull request merged**